### PR TITLE
[frontend] Cart api

### PR DESCRIPTION
**Pull request by UGisBusy**
Reeeeeeeeeeeeesolves #94 

### Changes
- [x] [/buyer/cart](http://localhost:8080/buyer/cart)
- cart page apis

#### Function
- carts section
  - change quantity of product from 0 to that stock of product
  - beware of that if quantity is equal 0 it will remove that row entirely and remove whole cart if there's no product in it
  - if there's any product in the cart is not availabe, it will alert and tell u what's went wrong
- checkout section (the canva)
  - see the detailed data
  - add/remove used coupons
  - select credit card
  - pay. will check if any card is selected and when succeed, link back to carts
- useable coupon section (the modal)
  - see the usable coupons
  - click one to use it 

### Testing Approach (if unconventional)
```sh
docker compose up --build
```
- login as customer(user)
- click and add product to cart in home page or from shop
- go to [/buyer/cart](http://localhost:8080/buyer/cart)
- try to checkout, add/remove usable coupon
- try to pay the cart
- go to [order history](http://localhost:8080/user/buyer/order) to see if the cart u paid is there

### CODE REVIEW

File: frontend/src/components/Cart.tsx [hunk 1]
```diff
@@ -363,13 +391,26 @@ const Cart = ({ data, onRefetch }: Props) => {
...
+                    />
+                  </div>
+                </Col>
+              ))
+            ) : (
+              <Col xs={12} className='center'>
+                <h2>No usable coupon 😢</h2>
               </Col>
-            ))}
+            )}
```
Comment (kryr34) [inline frontend/src/components/Cart.tsx]: why isn't `usableCouponData === undefined` showing 😢

File: frontend/src/components/Cart.tsx [hunk 2]
```diff
@@ -86,123 +85,165 @@ const ContentStyle = {
...
-    onRefetchUsableCoupon();
-    setModalShow(true);
+    const resp = await fetch(`/api/buyer/cart/${cartInfo.id}/checkout`, {
+      method: 'POST',
+      headers: {
+        Authorization: `Bearer ${token}`,
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(checkoutData?.payments[watch('card_id')]),
```
Comment (kryr34) [inline frontend/src/components/Cart.tsx]: is POST null JSON expected behavior?

**Review by kryr34 — COMMENTED** [review]
LGTM w/o test

File: frontend/src/components/Cart.tsx (hunk 1 above)
Comment (UGisBusy) [reply frontend/src/components/Cart.tsx]: uhh that is too make sure usableCouponData is array brefore checking sableCouponData.length

Comment (kryr34) [reply frontend/src/components/Cart.tsx]: I thought `usableCouponData?.` is enough

File: frontend/src/components/Cart.tsx (hunk 2 above)
Comment (UGisBusy) [reply frontend/src/components/Cart.tsx]: no, and i think ill add a comment explan this

File: frontend/src/components/Cart.tsx (hunk 1 above)
Comment (kryr34) [reply frontend/src/components/Cart.tsx]: > uhh that is too make sure usableCouponData is array brefore checking sableCouponData.length

wait shouldn't it be `usableCouponData === undefined && usableCouponData?.length > 0` ?

**Comment by UGisBusy** [comment]
## Why every ``checkoutData`` can be null/undefined ?
the ``checkoutData`` is the result of fetching buyer get checkout api.
 it only calls when try to see the checkout detail (the canva).
this means when u first enter the page, the value of  ``checkoutData``  is undefined.

same thing as above but more code-like:
the useQuery hook for checkout api is set to ``enabled = false``.
so it idles (at the status of loading) untill u call refetch() and the data before that was undefined.

## Can this be a problem?
when the fetch failed, i block the result from showing by:
```ts
  if (checkoutStatus === 'error' && canvaShow) {
    setCanvaShow(false);
  }
  if (usableCouponStatus === 'error' && modalShow) {
    setModalShow(false);
  }
```

what if i try do stuff with ``checkoutData`` before fetch?
no u cant. 
everything useing ``checkoutData`` is behind the canva, and the canva only show if it successfully fetch.

File: frontend/src/components/Cart.tsx (hunk 2 above)
Comment (kryr34) [reply frontend/src/components/Cart.tsx]: > no

then I think you should check `if checkoutData === null or undefined` before the POST.
if adding `?` everywhere is a JS convention, then I have no objections.

Comment (UGisBusy) [reply frontend/src/components/Cart.tsx]: yes im doing that rn
but not the element part tho

File: frontend/src/components/Cart.tsx (hunk 1 above)
Comment (NOOBDY) [reply frontend/src/components/Cart.tsx]: shouldn't it be
```ts
usableCouponData && usableCouponData.length > 0
```
or
```ts
Array.isArray(usableCouponData) && usableCouponData.length > 0
```

File: frontend/src/components/Cart.tsx [hunk 3]
```diff
@@ -159,17 +159,24 @@ const Cart = ({ products, cartInfo, refresh }: Props) => {
...
-    setModalShow(true);
+    if (usableCouponStatus === 'success') {
+      setModalShow(true);
+    }
   };
 
   const onPay = async () => {
+    if (checkoutData === undefined || checkoutStatus !== 'success') {
+      return;
+    }
```
Comment (kryr34) [inline frontend/src/components/Cart.tsx]: doesn't this need `alert` or smth?

File: frontend/src/components/Cart.tsx (hunk 1 above)
Comment (Error0229) [reply frontend/src/components/Cart.tsx]: UG said it already fixed

**Review by kryr34 — DISMISSED** [review]
LGTM w/o test (idk if you need my approve, but I think this should be merged after at least 2 approve from someone who really knows React. but I still approved this, just to avoid someone too busy to review)

**Review by Error0229 — APPROVED** [review]
Except 我們的coupon若是在台灣時間0~8點新增，且開始日期設為Today，他要到8點才能開始使用。 Everything works on my machine

**Review by NOOBDY — APPROVED** [review]

**Pull request merged**